src/jshint.js

```diff
@@ -5032,115 +4831,625 @@ var JSHINT = (function() {
 
       atTop: function() {
-        return _variables.length === 1;
+        return _scopeStack.length === 1;
+      },
+
+      /**
+       * for the exported options, indicating a variable is used outside the file
+       */
+      addExported: _addUsage,
+
+      /**
+       * Mark an indentifier as es6 module exported
+       */
+      setExported: function(labelName, token) {
+        this.current.use(labelName, token);
       },
 
-      setExported: function(id) {
-        if (funct["(blockscope)"].atTop()) {
-          var item = _current[id];
-          if (item && item["(token)"]) {
-            item["(token)"].exported = true;
+      functionScope: {
+        /**
+         * Returns the label type given certain options
+         * @param labelName
+         * @param {Object=} options
+         * @param {Boolean=} options.onlyBlockscoped - only include block scoped labels
+         * @param {Boolean=} options.excludeParams - exclude the param scope
+         * @param {Boolean=} options.excludeCurrent - exclude the current scope
+         */
+        labeltype: function(labelName, options) {
+          var onlyBlockscoped = options && options.onlyBlockscoped;
+          var excludeParams = options && options.excludeParams;
+          var currentScopeIndex = _scopeStack.length - (options && options.excludeCurrent ? 2 : 1);
+          for (var i = currentScopeIndex; i >= 0; i--) {
+            var current = _scopeStack[i];
+            if (_.has(current["(labels)"], labelName) &&
+              (!onlyBlockscoped || current["(labels)"][labelName]["(blockscoped)"])) {
+              return current["(labels)"][labelName]["(type)"];
+            }
+            var scopeCheck = excludeParams ? _scopeStack[ i - 1 ] : current;
+            if (scopeCheck && scopeCheck["(isParams)"] === "function") {
+              return null;
+            }
+          }
+          return null;
+        },
+        hasBreakLabel: function(labelName) {
+          for (var i = _scopeStack.length - 1; i >= 0; i--) {
+            var current = _scopeStack[i];
+
+            if (_.has(current["(breakLabels)"], labelName)) {
+              return true;
+            }
+            if (current["(isParams)"] === "function") {
+              return false;
+            }
           }
+          return false;
+        },
+        /**
+         * Returns if the label is in the current function scope
+         * See functionScope.labelType for options
+         */
+        has: function(labelName, options) {
+          return Boolean(this.labeltype(labelName, options));
+        },
+
+        /**
+         * Adds a new function scoped variable
+         * see current.add for block scoped
+         */
+        add: function(labelName, type, tok, unused) {
+          _current["(labels)"][labelName] = {
+            "(type)" : type,
+            "(token)": tok,
+            "(blockscoped)": false,
+            "(unused)": unused };
         }
       },
 
       current: {
```

Comment:
The name `current` makes sense--`funct["scope"].current` accurately describes the object as the current scope. The `functionScope` property muddies this a little bit--given `funct["scope"].current`, it's not so clear that `funct["scope"].functionScope` is the *current* function scope... or that `current` is a block scope specifically.

What do you think about

- `funct["scope"].current` -> `funct["scope"].currentBlock`
- `funct["scope"].functionScope` -> `funct["scope"].currentFunct`

or (since that's a lot of typing):

- `funct["scope"].current` -> `funct["scope"].block`
- `funct["scope"].functionScope` -> `funct["scope"].funct`

?